src/dmd/statement.d

```diff
@@ -449,6 +449,89 @@ extern (C++) abstract class Statement : ASTNode
     inout(CompoundDeclarationStatement) isCompoundDeclarationStatement() { return stmt == STMT.CompoundDeclaration ? cast(typeof(return))this : null; }
 }
 
+
+// TODO Make this function iterative if it turns out to be a bottleneck
+private Statement stripScopeStatment(Statement s)
+{
+    Statement result = s;
+
+    if (s)
+    {
+        ScopeStatement ss = s.isScopeStatement();
+        while(ss && ss.statement)
+        {
+            result = ss.statement;
+            ss = result.isScopeStatement();
+        }
+    }
+
+    return result;
+}
+/**
+* Returns:
+*     Wheter ($D stmt) containts assert(ctfe)
+* Params:
+*  stmt = CompoundStatement to check
+*/
+bool containsAssertCtfe(CompoundStatement stmt)
+{
+    bool result = false;
+
+    Statements *statements;
+    if (stmt !is null)
+    {
+        if (CompoundStatement cs = stmt.isCompoundStatement())
+        {
+            statements = cs.statements;
+        }
+    }
+
+    if (statements !is null)
+    {
+        foreach(s;(*statements))
+        {
+            if (!s)
+                continue;
+
+            s = stripScopeStatment(s);
+            if (isAssertCtfe(s))
+            {
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+/**
+* Returns:
+*     Wheter ($D stmt) is assert(ctfe)
+* Params:
+*  stmt = Statement to check
+*/
+extern (C++) public bool isAssertCtfe(Statement s)
+{
+    if (s)
+    {
+        if (auto es = s.isExpStatement())
+        {
+            if (auto ae = (es.exp ? es.exp.isAssertExp() : null))
+            {
+                if (auto ve = (ae.e1 ? ae.e1.isVarExp() : null))
+                {
+                    if (ve && ve.var && ve.var.ident == Id.ctfe)
+                    {
```

Comment:
This many nested ifs can't be right.

Reply:
I matching a pattern on a tree.
The ifs accurately represent the decisions which have to be made in order to match.

What would you consider as an alternative?